src/preprocess/mathjax.rs

```diff
@@ -0,0 +1,261 @@
+            ([^$]+)  # followed by one or more things other than a dollar sign
+            (\$)     # followed by a closing dollar sign.
+
+            |        # or
+
+                     # Legacy inline mathematics
+            (\\\\\() # An escaped opening bracket `\\(`
+            ([^)]+)  # followed by one or more other things TODO provide correct regexp.
+            (\\\\\)) # followed by a closing bracket `\\)`
+
+            |        # or
+
+                     # Legacy block mathematics
+            (\\\\\[) # An escaped opening bracket `\\[`
+            ([^$]+)  # followed by one ore more other things TODO provide correct regexp.
+            (\\\\\]) # followed by a closing bracket `\\]`
+        ").unwrap();
+    }
+    MathematicsIterator(REGEXP.captures_iter(content))
+}
+
+struct MathematicsIterator<'a>(CaptureMatches<'a, 'a>);
+
+impl<'a> Iterator for MathematicsIterator<'a> {
+    type Item = Mathematics<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for capture in &mut self.0 {
+            if let mathematics @ Some(_) = Mathematics::from_capture(capture) {
+                return mathematics;
+            }
+        }
+        None
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Mathematics<'a> {
+    start_index: usize,
+    end_index: usize,
+    kind: Kind,
+    text: &'a str,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum Kind {
+    Inline,
+    Block,
+    LegacyInline,
+    LegacyBlock,
+}
+
+impl<'a> Mathematics<'a> {
+    fn from_capture(captures: Captures<'a>) -> Option<Self> {
+        let kind =
+            captures.get(1).or(captures.get(4)).or(captures.get(7)).or(captures.get(10))
+            .map(|delimiter|
+                 match delimiter.as_str() {
+                     "$$"    => Kind::Block,
+                     "$"     => Kind::Inline,
+                     "\\\\[" => Kind::LegacyBlock,
+                     _       => Kind::LegacyInline,
+                 })
+            .expect("captured mathematics should have opening delimiter at the provided indices");
+
+        captures.get(0).map(|m| Mathematics {
+            start_index: m.start(),
+            end_index: m.end(),
+            kind: kind,
+            text: kind.text(m.as_str()),
+        })
+    }
+
+    fn replacement(&self) -> String {
+        let mut replacement = String::new();
+        match self.kind {
+            Kind::Block  | Kind::LegacyBlock  => {
+                replacement.push_str("<div class=\"math\">$$");
+                replacement.push_str(self.text);
+                replacement.push_str("$$</div>");
+            },
+            Kind::Inline | Kind::LegacyInline => {
+                replacement.push_str("<span class=\"inline math\">$");
+                replacement.push_str(self.text);
+                replacement.push_str("$</span>");
+            },
+        }
+        replacement
+    }
+}
+
+impl Kind {
+    fn text<'a>(&self, delimited_text: &'a str) -> &'a str {
```

Comment:
I would have thought this method belongs on the `Mathematics` type.

Reply:
I placed it on `Kind` because I needed the body of the delimited text when I construct a `Mathematics` structure. This could well be done in a constructor.

Reply:
I opted for  a auxiliary function with the same functionality. :heavy_check_mark: